src/librados/IoCtxImpl.cc

```diff
@@ -2020,9 +2022,19 @@ void librados::IoCtxImpl::C_aio_Complete::finish(int r)
   c->cond.Signal();
 
   if (r == 0 && c->blp && c->blp->length() > 0) {
-    if (c->out_buf && !c->blp->is_provided_buffer(c->out_buf))
-      c->blp->copy(0, c->blp->length(), c->out_buf);
```

Comment:
instead of copying all the content from `c->blp` to `c->out_buf`, i think we should take this as an indication of the ERANGE: the buffer provided by caller is shorter than the read op's returned payload. in this case, we should just return -ERANGE. as both simple and async messengers *append* new memory chunks to the provided bufferlist if it's smaller than `header.data_len`. in that case, `c->blp->is_provided_buffer(c->out_buf)` would evaluate to false. and the first segment in `c->blp` should always be `c->out_buf`. so i think `c->blp->contains_buffer(c->out_buf)` will always be true in this case, unless we **change** the bufferlist provided by the `aio_execute()`'s caller in a way that this bufferlist's first element is changed. but i don't think it's possible.